pkg/reconciler/knativeserving/ingress/ingress_test.go

```diff
@@ -0,0 +1,414 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"context"
+	"os"
+	"testing"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	mf "github.com/manifestival/manifestival"
+	servingv1alpha1 "knative.dev/operator/pkg/apis/operator/v1alpha1"
+	"knative.dev/operator/pkg/reconciler/common"
+	util "knative.dev/operator/pkg/reconciler/common/testing"
+)
+
+const NumberIngressResource = 5
+
+func TestGetIngress(t *testing.T) {
+	os.Setenv(common.KoEnvKey, "testdata/kodata")
+	defer os.Unsetenv(common.KoEnvKey)
+
+	tests := []struct {
+		name                 string
+		targetVersion        string
+		expected             bool
+		expectedIngressesNum int
+	}{{
+		name:                 "Available ingresses",
+		targetVersion:        "0.18",
+		expected:             true,
+		expectedIngressesNum: NumberIngressResource,
+	}, {
+		name:                 "Unavailable ingresses",
+		targetVersion:        "0.17",
+		expected:             false,
+		expectedIngressesNum: 0,
+	}, {
+		name:                 "Missing version",
+		targetVersion:        "",
+		expected:             true,
+		expectedIngressesNum: 0,
+	}}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			manifest, _ := mf.ManifestFrom(mf.Slice{})
+			err := getIngress(tt.targetVersion, &manifest)
+			util.AssertEqual(t, err == nil, tt.expected)
+			util.AssertEqual(t, len(manifest.Resources()), tt.expectedIngressesNum)
+		})
+	}
+}
+
+func TestAppendInstalledIngresses(t *testing.T) {
+	os.Setenv(common.KoEnvKey, "testdata/kodata")
+	defer os.Unsetenv(common.KoEnvKey)
+
+	tests := []struct {
+		name                 string
+		instance             servingv1alpha1.KnativeServing
+		expected             bool
+		expectedIngressesNum int
```

Comment:
I don't quite understand what this checks. Can we make the test explicitly check that only istio is loaded by default, Kourier if specified and Istio AND Kourier if both are specified?

Reply:
This test cases check whether it can get exactly the same number of resources available in the ingress directory. I change the name into `expectedResourcesNum`.
To verify if several specified ingresses are loaded, I will add them into the test case `TestTransformers`.